mmdeploy/apis/onnx/export.py

```diff
@@ -0,0 +1,127 @@
+        >>>     'place/to/save/model',
+        >>>     backend='tensorrt',
+        >>>     input_names=['input'],
+        >>>     output_names=['output'],
+        >>>     dynamic_axes={'input': {
+        >>>         0: 'batch',
+        >>>         2: 'height',
+        >>>         3: 'width'
+        >>>     }})
+
+    Args:
+        model (torch.nn.Module): the model to be exported.
+        args (torch.Tensor|Tuple|Dict): Dummy input of the model.
+        output_path_prefix (str): The output file prefix. The model will
+            be saved to `<output_path_prefix>.onnx`.
+        backend (Backend|str): Which backend will the graph be used. Different
+            backend would generate different graph.
+        input_metas (Dict): The constant inputs of the model.
+        context_info (Dict): The information that would be used in the context
+            of exporting.
+        input_names (Sequence[str]): The input names of the model.
+        output_names (Sequence[str]): The output names of the model.
+        opset_version (int): The version of ONNX opset version. 11 as default.
+        dynamic_axes (Dict): The information used to determine which axes are
+            dynamic.
+        verbose (bool): Enable verbose model on `torch.onnx.export`.
+        keep_initializers_as_inputs (bool): Whether we should add inputs for
+            each initializer.
+    """
+    output_path = output_path_prefix + '.onnx'
+
+    logger = get_root_logger()
+    logger.info(f'Export PyTorch model to ONNX: {output_path}.')
+
+    def _add_or_update(cfg: dict, key: str, val: Any):
+        if key in cfg and isinstance(cfg[key], dict) and isinstance(val, dict):
+            cfg[key].update(val)
+        else:
+            cfg[key] = val
+
+    context_info = deepcopy(context_info)
+    deploy_cfg = context_info.pop('deploy_cfg', dict())
+    ir_config = dict(
+        type='onnx',
+        input_names=input_names,
+        output_names=output_names,
+        opset_version=opset_version,
+        dynamic_axes=dynamic_axes,
+        verbose=verbose,
+        keep_initializers_as_inputs=keep_initializers_as_inputs)
+    _add_or_update(deploy_cfg, 'ir_config', ir_config)
+
+    if isinstance(backend, Backend):
+        backend = backend.value
+    backend_config = dict(type=backend)
+    _add_or_update(deploy_cfg, 'backend_config', backend_config)
+
+    context_info['cfg'] = deploy_cfg
+    if 'backend' not in context_info:
+        context_info['backend'] = backend
+    if 'opset' not in context_info:
+        context_info['opset'] = opset_version
+
+    # patch model
+    patched_model = patch_model(model, cfg=deploy_cfg, backend=backend)
+
+    with RewriterContext(**context_info), torch.no_grad():
+        # patch input_metas
+        if input_metas is not None:
+            assert isinstance(
+                input_metas, dict
+            ), f'Expect input_metas type is dict, get {type(input_metas)}.'
+            model_forward = model.forward
```

Comment:
I guess that `model_forward` here is used to temporarily hold the function and `partial` works like `std::bind`, can you give me an example of  $MODEL_CONFIG and  $MODEL_PATH ?

Reply:
forward of some models in each repo accept non-tensor input, For example:
https://github.com/open-mmlab/mmdetection/blob/240d7a31c745578aa8c4df54c3074ce78b690c34/mmdet/models/detectors/base.py#L157

And it is not allowed in PyTorch ONNX exporter. Which only accepts Tesnor or Tuple of Tensor as the tracing parameters.
https://github.com/pytorch/pytorch/blob/18d46ea9fd5071820cdbe7595b99396a9153f716/torch/onnx/__init__.py#L84

Actually, this is not what I expected. `meta` could be added by `patch_model`, which can create a wrap of the given model. But since this refactor is for API not its implementation, can we ignore it for a while?

Reply:
You can use any config in MMDetection as the example.